Module name lookup optimizations

slavapestov opened the pull request:
Follow-up to https://github.com/apple/swift/pull/26626.

We have some quadratic behavior from repeatedly having to visit all SourceFiles in a ModuleDecl while doing name lookup, because some stuff was cached at the level of a SourceFile. By moving the caching up a level to ModuleDecl we avoid the iterations.

slavapestov commented:
I'm still working on some scale-tests, but this is ready for a preliminary review at least.

DougGregor commented on lib/AST/DeclContext.cpp:
```diff
@@ -475,6 +475,25 @@ unsigned DeclContext::getSemanticDepth() const {
+      if (PD->getAttrs().hasAttribute<ObjCAttr>())
+        return true;
+  } else if (auto *ED = dyn_cast<ExtensionDecl>(this)) {
+    if (auto *CD = dyn_cast<ClassDecl>(ED->getExtendedNominal()))
```
Please use `dyn_cast_or_null` on the result of `getExtendedNominal`, since it could fail. Alternatively, `getSelfClassDecl()` would let you combine the "class" and "extension of class" cases and probably needs less code.

slavapestov replied:
Alright, I'll fix that in the other PR (this one needs a bit more debugging -- the final commit doesn't pass tests)

davidungar commented on lib/AST/Decl.cpp:
```diff
@@ -2672,26 +2672,14 @@ bool ValueDecl::canBeAccessedByDynamicLookup() const {
-  // Dynamic lookup cannot find results within a non-protocol generic context,
-  // because there is no sensible way to infer the generic arguments.
-  if (getDeclContext()->isGenericContext() && !isa<ProtocolDecl>(nominalDC))
+  auto *dc = getDeclContext();
```
A real nit, and feel free to ignore: since `dc` isn't used again, I find it clearer to just write the `getDeclContext` call in the next line and not introduce a variable.

davidungar commented on lib/AST/DeclContext.cpp:
```diff
@@ -475,6 +475,25 @@ unsigned DeclContext::getSemanticDepth() const {
+bool DeclContext::mayContainMembersAccessedByDynamicLookup() const {
+  // Dynamic lookup can only find class and protocol members, or extensions of
+  // classes.
+  if (auto *NTD = dyn_cast<NominalTypeDecl>(this)) {
```
In reviewing this code, I wanted to read this function and ensure it was doing what the comment implied. But it defeats me, in my code-reviewing mode. Could you rewrite or refactor it to make it clearer? For instance, the flow can reach 485 if it's a ClassDecl, so why check for ProtocolDecl?

Would it work to just start with a check `isGenericContext` and return false if it is generic?

slavapestov replied:
Which part is unclear? Only one of line 482 or line 485 will test true. We're checking for class members, protocol members, and finally class extension members.

davidungar replied:
We discussed this just now offline. Your revised version (future PR?) is great

davidungar commented on lib/AST/DeclContext.cpp:
```diff
@@ -794,13 +813,19 @@ void IterableDeclContext::setMemberLoader(LazyMemberLoader *loader,
+    auto *IDC = const_cast<IterableDeclContext *>(this);
+    ctx.parseMembers(IDC);
+    IDC->HasUnparsedMembers = 0;
+  }
```
Nice and clear!

davidungar commented on lib/AST/Module.cpp:
```diff
@@ -186,58 +187,92 @@ SourceLookupCache &SourceFile::getCache() const {
-    if (auto *VD = dyn_cast<ValueDecl>(D))
+    if (auto *VD = dyn_cast<ValueDecl>(D)) {
       if (onlyOperators ? VD->isOperator() : VD->hasName()) {
         // Cache the value under both its compound name and its full name.
```
Not your change, but why? Why cache it under both names?? More explanation would help me.

slavapestov replied:
You can look up an operator as `+` or `+(_:_:)`

davidungar commented on lib/AST/Module.cpp:
```diff
@@ -118,7 +119,7 @@ BuiltinUnit::BuiltinUnit(ModuleDecl &M)
 SourceFile::~SourceFile() = default;
 
-class SourceFile::LookupCache {
+class swift::SourceLookupCache {
```
Could you add a comment for this class, explaining the "why"? Tx.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

davidungar reviewed (approved):
Overall, looks pretty good. I commented on a few places where things could be a bit clearer. I don't have a deep understanding of (in particular) the use of the cache in Module.cpp, so I could have missed some bugs.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

jrose-apple commented:
This seems suspect. What happens with private declarations? What about imports differing from file to file?

slavapestov commented:
@jrose-apple Which part is suspect? The per-SourceFile operations remain unchanged, this just adds a cache in front of the ModuleDecl implementations which use the FORWARD macro to iterate over all source files. The latter are called all over the place in name lookup. There should not be any semantic behavior changes here.

jrose-apple commented on lib/AST/Module.cpp:
```diff
@@ -407,15 +426,41 @@ void ModuleDecl::removeFile(FileUnit &existingFile) {
+  // FIXME: If we ever get mixed modules that contain both SourceFiles and other
+  // kinds of file units, this will break; there all callers of this function should
+  // themselves assert that all file units in the module are SourceFiles when this
+  // function returns true.
```
We do have this with SIB files. :-/ I guess if it's not tested it's not interesting, but…

slavapestov replied:
Yeah, that’s actually I’d tested, it’s why I bail out of the first file in SIL (if there’s a SIL file, there’s only one, and it’s the first). That deserves a comment.

jrose-apple replied:
You can (in theory) have a plain Swift file mixed with SIB files.

jrose-apple commented on lib/AST/Module.cpp:
```diff
@@ -1069,14 +1163,46 @@ SourceFile::getImportedModules(SmallVectorImpl<ModuleDecl::ImportedModule> &modu
-    if (filter.contains(requiredKind))
+    if (filter.contains(requiredKind)) {
+      // Filter out duplicate imports; they have no semantic effect.
+      auto iter = std::find_if(modules.begin(), modules.end(),
```
Nitpick: `llvm::find_if` (and `llvm::copy` elsewhere).

slavapestov replied:
Sure. What’s the difference?

jrose-apple replied:
These two are just easier to read (since you don't need to pass the iterators separately), but some of the other STL covers do extra checking (e.g. to make sure you aren't relying on iteration order for unordered containers).

jrose-apple commented:
Oops, I thought the existing caches were for the results of top-level lookup, not for just the decls in the SourceFile. Objection retracted (though that might still be a more useful place to cache things).

slavapestov commented:
@jrose-apple Yeah, access control and stuff like that happens at the level above this. Arguably imports should still be per-file and not per-module, but that's not how things work today...

jrose-apple commented:
Imports *are* per-file…when not `@_exported`.

jrose-apple commented:
(The ImportFilter really matters. Higher-level APIs translate queries into the correct ImportFilter.)

slavapestov commented on include/swift/AST/Module.h:
```diff
@@ -378,6 +401,12 @@ class ModuleDecl : public DeclContext, public TypeDecl {
                           VisibleDeclConsumer &Consumer,
                           NLKind LookupKind) const;
 
+  /// This is a hack for 'main' file parsing and the integrated REPL.
```
Because even though the REPL parses declarations in their own source file before adding them to the “combined” file, we still appear to do name lookups in the combined file, so we need a way to clear the cache.

slavapestov replied:
I’m actually still debugging some of this cache invalidation stuff. Ideally it would not be needed but we do lots of tricky things still.

slavapestov committed:
AST: Add per-module lookup cache

These operations are called frequently, and would get rather
expensive when a ModuleDecl contains a large number of
SourceFiles:

- lookupValue()
- lookupVisibleDecls()
- lookupClassMember()
- lookupClassMembers()

Add fast paths to handle the case where all file units in the
module are SourceFiles, by refactoring SourceFile::LookupCache
from a per-file to a per-module cache.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance
